x/sync/manager.go

```diff
@@ -63,6 +64,71 @@ const (
 	retryPriority
 )
 
+var _ DBClient = (*merkleDBSyncClient)(nil)
+
+type ClientConfig struct {
+	Hasher       merkledb.Hasher
+	BranchFactor merkledb.BranchFactor
+}
+
+type merkleDBSyncClient struct {
+	db        DB
+	config    *ClientConfig
+	tokenSize int
+	err       *utils.Atomic[error]
+	errorOnce sync.Once
+}
+
+func NewClient(db DB, config *ClientConfig) (*merkleDBSyncClient, error) {
+	if err := config.BranchFactor.Valid(); err != nil {
+		return nil, err
+	}
+
+	if config.Hasher == nil {
+		config.Hasher = merkledb.DefaultHasher
+	}
+
+	return &merkleDBSyncClient{
+		db:        db,
+		config:    config,
+		tokenSize: merkledb.BranchFactorToTokenSize[config.BranchFactor],
+		err:       utils.NewAtomic[error](nil),
+	}, nil
+}
+
+func (c *merkleDBSyncClient) Error() error {
+	return c.err.Get()
+}
+
+func (c *merkleDBSyncClient) setError(err error) {
+	c.errorOnce.Do(func() {
+		c.err.Set(err)
+		_ = c.db.Clear() // only report the original error
+	})
+}
+
+// FinalizeSync implements DBClient.
+func (c *merkleDBSyncClient) FinalizeSync(ctx context.Context, root ids.ID) error {
+	// If the root is empty, we can clear the database and finish immediately.
+	if root == ids.Empty {
+		if err := c.db.Clear(); err != nil {
+			c.setError(err)
+			return fmt.Errorf("failed to clear database: %w", err)
+		}
+		return nil
+	}
```

Comment:
The caller knows the root id... so why do we couple the behavior of clearing to the verification of the sync being done? We could just have two separate i/f signatures to check that the root is correct vs to skip syncing.